Public/Src/Engine/Scheduler/Fingerprints/ObservedInputProcessor.cs

```diff
@@ -892,11 +891,11 @@ private static bool IsPathWithinAny(PathTable pathTable, AbsolutePath path, IEnu
             where TTarget : struct, IObservedInputProcessingTarget<TObservation>
         {
             using (var pooledPathSet = Pools.GetAbsolutePathSet())
-            using (var pooledStringIdSet = Pools.GetStringIdSet())
             {
-                HashSet<StringId> accessedFileNames = pooledStringIdSet.Instance;
                 HashSet<AbsolutePath> visitedPaths = pooledPathSet.Instance;
 
+                HashSet<StringId> accessedFileNames = new HashSet<StringId>(pathTable.StringTable.CaseInsensitiveEqualityComparer);

```

Comment:
That'd be awesome if you try to use the pooled HashSet objects. You can add that to ObservedInputProcessingState:

```
        private static volatile ObjectPool<ObservedInputProcessingState> s_pool;
        public readonly HashSet<StringId> AccessedFileNames;
        private readonly IEqualityComparer<StringId> m_comparer;
        private ObservedInputProcessingState(IEqualityComparer<StringId> comparer)
        {
            m_comparer = comparer;
            AccessedFileNames = new HashSet<StringId>(comparer);
        }
        public static ObservedInputProcessingState GetInstance(IEqualityComparer<StringId> comparer)
        {
            if (s_pool == null)
            {
                s_pool = new ObjectPool<ObservedInputProcessingState>(
                    () => new ObservedInputProcessingState(comparer),
                    state => state.Clear());
            }

            return s_pool.GetInstance().Instance;
        }
```

In theory, this is not fully thread-safe as you can overwrite s_pool variable even though s_pool is volatile. However, that's still perfectly fine.

Reply:
I thought about doing this (though in a little bit different way, i.e., by maintaining a map from pathtable/comparer to a HashSet pool); however, pooling hashsets won't give us any perf benefits because at the end of the day we are still creating a new hashset in `SearchPathDirectoryMembershipFilter` method. So instead of making it there, I moved the logic to this method and stopped using pooled `stringid` hashset

---
In reply to: [348068346](https://github.com/microsoft/BuildXL/pull/1236#discussion_r348068346) [](ancestors = 348068346)